File: packages/shared-contracts/src/entities.ts
import type { PrivacyDataClass } from "./privacy";

export const ENTITY_TYPES = [
  "habit",
  "habit_log",
  "task",
  "routine",
  "artifact",
  "conversation",
  "experiment",
  "calendar_block",
  "day",
  "time_window",
] as const;

export type EntityType = (typeof ENTITY_TYPES)[number];

export const LAYER_0_ENTITY_TYPES = [
  "habit",
  "habit_log",
  "task",
  "routine",
  "artifact",
  "conversation",
  "calendar_block",
  "day",
  "time_window",
] as const;

export type Layer0EntityType = (typeof LAYER_0_ENTITY_TYPES)[number];

/** Incoming aliases. Stored and returned refs always use the canonical type. */
export const ENTITY_TYPE_ALIASES = {
  report: "artifact",
  calendar: "calendar_block",
} as const;

export type EntityTypeAlias = keyof typeof ENTITY_TYPE_ALIASES;

export type EntityRef = {
  type: EntityType;
  id: string;
};

export type EntityAvailability = "ok" | "unknown" | "deleted" | "forbidden";

export type EntitySummary = {
  ref: EntityRef;
  title: string;
  subtitle?: string;
  status?: string;
  icon?: string;
  route: string;
  updatedAt?: string;
  privacyClass: string;
  availability: EntityAvailability;
};

export type RelatedEntitySource = "fk" | "artifact_link" | "authored";

export type RelatedEntity = {
  ref: EntityRef;
  relationship: string;
  source: RelatedEntitySource;
};

export const AUTHORED_RELATIONSHIPS = [
  "references",
  "mentions",
  "supports",
  "contradicts",
  "evidence_for",
] as const;

export type AuthoredRelationship = (typeof AUTHORED_RELATIONSHIPS)[number];

export const ENTITY_MENTION_TOKEN_PATTERN = /\[\[([a-z_]+):([^\]]+)\]\]/g;

const ISO_DAY = /^\d{4}-\d{2}-\d{2}$/;
const ISO_WINDOW = /^(\d{4}-\d{2}-\d{2})\/(\d{4}-\d{2}-\d{2})$/;

const ENTITY_PRIVACY_CLASS: Record<EntityType, PrivacyDataClass> = {
  habit: "habit_definition",
  habit_log: "habit_log",
  task: "task",
  routine: "routine",
  artifact: "ai_content",
  conversation: "ai_content",
  experiment: "task",
  calendar_block: "task",
  day: "habit_log",
  time_window: "habit_log",
};

export function canonicalEntityType(value: unknown): EntityType | null {
  if (typeof value !== "string") return null;
  if ((ENTITY_TYPES as readonly string[]).includes(value)) return value as EntityType;
  if (value in ENTITY_TYPE_ALIASES) {
    return ENTITY_TYPE_ALIASES[value as EntityTypeAlias];
  }
  return null;
}

export function isEntityType(value: unknown): value is EntityType {
  return typeof value === "string" && (ENTITY_TYPES as readonly string[]).includes(value);
}

export function entityTypeToPrivacyClass(type: EntityType): PrivacyDataClass {
  const canonical = canonicalEntityType(type) || type;
  return ENTITY_PRIVACY_CLASS[canonical];
}

export function entityRefKey(ref: EntityRef): string {
  const type = canonicalEntityType(ref.type) || ref.type;
  return `${type}:${ref.id}`;
}

export function entityRoute(type: EntityType, id: string): string {
  const encoded = encodeURIComponent(id);
  const canonical = canonicalEntityType(type) || type;
  switch (canonical) {
    case "habit":
      return `/dashboard?view=metrics&habit=${encoded}`;
    case "habit_log":
      return `/activity?logId=${encoded}`;
    case "task":
      return `/tasks?task=${encoded}`;
    case "routine":
      return `/routines?routine=${encoded}`;
    case "artifact":
      return `/reports?artifactId=${encoded}`;
    case "conversation":
      return `/chat?conversation=${encoded}`;
    case "experiment":
      return `/experiments/${encoded}`;
    case "calendar_block":
      return `/calendar?block=${encoded}`;
    case "day":
      return `/calendar?date=${encoded}`;
    case "time_window": {
      const [from, to] = id.split("/");
      const start = encodeURIComponent(from || id);
      const end = encodeURIComponent(to || from || id);
      return `/activity?from=${start}&to=${end}`;
    }
  }
}

export function unavailableEntitySummary(
  ref: EntityRef,
  availability: Exclude<EntityAvailability, "ok">,
): EntitySummary {
  const title =
    availability === "forbidden"
      ? "Unavailable"
      : availability === "deleted"
        ? "Deleted"
        : "Unknown";
  return {
    ref,
    title,
    route: entityRoute(ref.type, ref.id),
    privacyClass: entityTypeToPrivacyClass(ref.type),
    availability,
  };
}

export function formatEntityMentionToken(ref: EntityRef): string {
  const type = canonicalEntityType(ref.type) || ref.type;
  return `[[${type}:${ref.id}]]`;
}

export function parseEntityMentionTokens(text: string): EntityRef[] {
  const items: EntityRef[] = [];
  const seen = new Set<string>();
  for (const part of splitEntityMentionText(text)) {
    if (part.kind !== "mention") continue;
    const key = entityRefKey(part.ref);
    if (seen.has(key)) continue;
    seen.add(key);
    items.push(part.ref);
  }
  return items;
}

export type EntityMentionSegment =
  | { kind: "text"; value: string }
  | { kind: "mention"; ref: EntityRef }
  | { kind: "unknown"; raw: string };

export function splitEntityMentionText(text: string): EntityMentionSegment[] {
  const source = String(text || "");
  const parts: EntityMentionSegment[] = [];
  const pattern = new RegExp(ENTITY_MENTION_TOKEN_PATTERN.source, "g");
  let lastIndex = 0;
  for (const match of source.matchAll(pattern)) {
    const index = match.index ?? 0;
    if (index > lastIndex) {
      parts.push({ kind: "text", value: source.slice(lastIndex, index) });
    }
    const type = canonicalEntityType(match[1]);
    const id = (match[2] || "").trim();
    if (type && id) {
      parts.push({ kind: "mention", ref: { type, id } });
    } else {
      parts.push({ kind: "unknown", raw: match[0] });
    }
    lastIndex = index + match[0].length;
  }
  if (lastIndex < source.length) {
    parts.push({ kind: "text", value: source.slice(lastIndex) });
  }
  return parts;
}

export function insertEntityMentionToken(text: string, ref: EntityRef): string {
  const token = formatEntityMentionToken(ref);
  const replaced = String(text || "").replace(/(^|\s)@([^\s@]*)$/, `$1${token} `);
  if (replaced !== text) return replaced;
  const trimmed = String(text || "").replace(/\s+$/, "");
  return trimmed ? `${trimmed} ${token} ` : `${token} `;
}

export function stripEntityMentionTokens(text: string, titles?: Map<string, string>): string {
  return String(text || "").replace(ENTITY_MENTION_TOKEN_PATTERN, (_all, rawType, rawId) => {
    const type = canonicalEntityType(rawType);
    const id = String(rawId || "").trim();
    if (!type || !id) return "";
    return titles?.get(`${type}:${id}`) || id;
  });
}

export function isDayId(id: string): boolean {
  return ISO_DAY.test(id);
}

export function isTimeWindowId(id: string): boolean {
  const match = id.match(ISO_WINDOW);
  return Boolean(match && match[1] <= match[2]);
}

export type ParsedDateMention =
  | { type: "day"; id: string }
  | { type: "time_window"; id: string };

function formatLocalDay(date: Date): string {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const day = String(date.getDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

function startOfLocalWeek(date: Date): Date {
  const next = new Date(date);
  next.setHours(12, 0, 0, 0);
  next.setDate(next.getDate() - next.getDay());
  return next;
}

export function parseDateMentionQuery(query: string, now: Date = new Date()): ParsedDateMention | null {
  const raw = query.trim().toLowerCase();
  if (!raw) return null;
  if (ISO_DAY.test(raw)) return { type: "day", id: raw };
  const windowMatch = raw.match(ISO_WINDOW);
  if (windowMatch && windowMatch[1] <= windowMatch[2]) {
    return { type: "time_window", id: `${windowMatch[1]}/${windowMatch[2]}` };
  }

  const today = new Date(now);
  today.setHours(12, 0, 0, 0);

  if (raw === "today") return { type: "day", id: formatLocalDay(today) };
  if (raw === "yesterday") {
    const date = new Date(today);
    date.setDate(date.getDate() - 1);
    return { type: "day", id: formatLocalDay(date) };
  }
  if (raw === "tomorrow") {
    const date = new Date(today);
    date.setDate(date.getDate() + 1);
    return { type: "day", id: formatLocalDay(date) };
  }
  if (raw === "this week") {
    const start = startOfLocalWeek(today);
    const end = new Date(start);
    end.setDate(end.getDate() + 6);
    return { type: "time_window", id: `${formatLocalDay(start)}/${formatLocalDay(end)}` };
  }
  if (raw === "last week") {
    const start = startOfLocalWeek(today);
    start.setDate(start.getDate() - 7);
    const end = new Date(start);
    end.setDate(end.getDate() + 6);
    return { type: "time_window", id: `${formatLocalDay(start)}/${formatLocalDay(end)}` };
  }
  if (raw === "last 7 days") {
    const start = new Date(today);
    start.setDate(start.getDate() - 6);
    return { type: "time_window", id: `${formatLocalDay(start)}/${formatLocalDay(today)}` };
  }
  return null;
}

export function virtualDateSummary(ref: ParsedDateMention): EntitySummary {
  if (ref.type === "day") {
    const date = new Date(`${ref.id}T12:00:00`);
    const valid = !Number.isNaN(date.getTime());
    return {
      ref: { type: "day", id: ref.id },
      title: valid
        ? date.toLocaleDateString(undefined, { month: "short", day: "numeric", year: "numeric" })
        : ref.id,
      subtitle: valid ? date.toLocaleDateString(undefined, { weekday: "long" }) : undefined,
      route: entityRoute("day", ref.id),
      privacyClass: entityTypeToPrivacyClass("day"),
      availability: valid ? "ok" : "unknown",
    };
  }
  const [from, to] = ref.id.split("/");
  return {
    ref: { type: "time_window", id: ref.id },
    title: from && to && from !== to ? `${from} – ${to}` : from || ref.id,
    subtitle: "Date range",
    route: entityRoute("time_window", ref.id),
    privacyClass: entityTypeToPrivacyClass("time_window"),
    availability: isTimeWindowId(ref.id) ? "ok" : "unknown",
  };
}

File: apps/backend/schemas/entities.py
"""Shared Entity Protocol contracts."""

from __future__ import annotations

import re
from datetime import date, timedelta
from typing import List, Literal, Optional, Tuple

from pydantic import BaseModel, Field, field_validator

from services.privacy_policy import data_class_for_entity_type


ENTITY_TYPES = (
    "habit",
    "habit_log",
    "task",
    "routine",
    "artifact",
    "conversation",
    "experiment",
    "calendar_event",
    "calendar_occurrence",
    "day",
    "time_window",
)

LAYER_0_ENTITY_TYPES = (
    "habit",
    "habit_log",
    "task",
    "routine",
    "artifact",
    "conversation",
    "calendar_event",
    "calendar_occurrence",
    "day",
    "time_window",
)

ENTITY_TYPE_ALIASES = {
    "report": "artifact",
    "calendar": "calendar_event",
}

ENTITY_MENTION_TOKEN_RE = re.compile(r"\[\[([a-z_]+):([^\]]+)\]\]")
ISO_DAY_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
ISO_WINDOW_RE = re.compile(r"^(\d{4}-\d{2}-\d{2})/(\d{4}-\d{2}-\d{2})$")

EntityType = Literal[
    "habit",
    "habit_log",
    "task",
    "routine",
    "artifact",
    "conversation",
    "experiment",
    "calendar_event",
    "calendar_occurrence",
    "day",
    "time_window",
]
EntityAvailability = Literal["ok", "unknown", "deleted", "forbidden"]
RelatedEntitySource = Literal["fk", "artifact_link", "authored"]
AuthoredRelationship = Literal["references", "mentions", "supports", "contradicts", "evidence_for"]
AuthoredProvenance = Literal["user", "assistant", "workflow"]


def canonical_entity_type(value: Optional[str]) -> Optional[str]:
    if not value:
        return None
    if value in ENTITY_TYPE_ALIASES:
        return ENTITY_TYPE_ALIASES[value]
    if value in ENTITY_TYPES:
        return value
    return None


def entity_route(entity_type: str, entity_id: str) -> str:
    encoded = entity_id
    canonical = canonical_entity_type(entity_type) or entity_type
    if canonical == "habit":
        return f"/dashboard?view=metrics&habit={encoded}"
    if canonical == "habit_log":
        return f"/activity?logId={encoded}"
    if canonical == "task":
        return f"/tasks?task={encoded}"
    if canonical == "routine":
        return f"/routines?routine={encoded}"
    if canonical == "artifact":
        return f"/reports?artifactId={encoded}"
    if canonical == "conversation":
        return f"/chat?conversation={encoded}"
    if canonical == "experiment":
        return f"/experiments/{encoded}"
    if canonical == "calendar_event":
        return f"/calendar?event={encoded}"
    if canonical == "calendar_occurrence":
        return f"/calendar?occurrence={encoded}"
    if canonical == "day":
        return f"/calendar?date={encoded}"
    if canonical == "time_window":
        parts = entity_id.split("/", 1)
        start = parts[0]
        end = parts[1] if len(parts) > 1 else parts[0]
        return f"/activity?from={start}&to={end}"
    return f"/{canonical}?id={encoded}"


def parse_entity_mention_tokens(text: Optional[str]) -> List["EntityRef"]:
    items: List[EntityRef] = []
    seen: set[Tuple[str, str]] = set()
    for match in ENTITY_MENTION_TOKEN_RE.finditer(text or ""):
        canonical = canonical_entity_type(match.group(1))
        entity_id = (match.group(2) or "").strip()
        if not canonical or not entity_id:
            continue
        key = (canonical, entity_id)
        if key in seen:
            continue
        seen.add(key)
        items.append(EntityRef(type=canonical, id=entity_id))
    return items


def is_day_id(value: str) -> bool:
    return bool(ISO_DAY_RE.match(value or ""))


def is_time_window_id(value: str) -> bool:
    match = ISO_WINDOW_RE.match(value or "")
    return bool(match and match.group(1) <= match.group(2))


def parse_date_mention_query(query: str, today: Optional[date] = None) -> Optional[Tuple[str, str]]:
    raw = (query or "").strip().lower()
    if not raw:
        return None
    if ISO_DAY_RE.match(raw):
        return ("day", raw)
    window = ISO_WINDOW_RE.match(raw)
    if window and window.group(1) <= window.group(2):
        return ("time_window", f"{window.group(1)}/{window.group(2)}")

    current = today or date.today()
    weekday = current.weekday()  # Monday=0; JS Date.getDay() is Sunday=0
    sunday_offset = (weekday + 1) % 7
    week_start = current - timedelta(days=sunday_offset)

    if raw == "today":
        return ("day", current.isoformat())
    if raw == "yesterday":
        return ("day", (current - timedelta(days=1)).isoformat())
    if raw == "tomorrow":
        return ("day", (current + timedelta(days=1)).isoformat())
    if raw == "this week":
        week_end = week_start + timedelta(days=6)
        return ("time_window", f"{week_start.isoformat()}/{week_end.isoformat()}")
    if raw == "last week":
        start = week_start - timedelta(days=7)
        end = start + timedelta(days=6)
        return ("time_window", f"{start.isoformat()}/{end.isoformat()}")
    if raw == "last 7 days":
        start = current - timedelta(days=6)
        return ("time_window", f"{start.isoformat()}/{current.isoformat()}")
    return None


def virtual_date_summary(entity_type: str, entity_id: str) -> Optional["EntitySummary"]:
    if entity_type == "day" and is_day_id(entity_id):
        try:
            parsed = date.fromisoformat(entity_id)
        except ValueError:
            return None
        return EntitySummary(
            ref=EntityRef(type="day", id=entity_id),
            title=f"{parsed.strftime('%b')} {parsed.day}, {parsed.year}",
            subtitle=parsed.strftime("%A"),
            route=entity_route("day", entity_id),
            privacyClass=data_class_for_entity_type("day"),
            availability="ok",
        )
    if entity_type == "time_window" and is_time_window_id(entity_id):
        start, end = entity_id.split("/", 1)
        title = start if start == end else f"{start} – {end}"
        return EntitySummary(
            ref=EntityRef(type="time_window", id=entity_id),
            title=title,
            subtitle="Date range",
            route=entity_route("time_window", entity_id),
            privacyClass=data_class_for_entity_type("time_window"),
            availability="ok",
        )
    return None


class EntityRef(BaseModel):
    type: EntityType
    id: str

    @field_validator("type", mode="before")
    @classmethod
    def canonicalize_type(cls, value: object) -> object:
        if not isinstance(value, str):
            return value
        canonical = canonical_entity_type(value)
        return canonical or value


class EntitySummary(BaseModel):
    ref: EntityRef
    title: str
    subtitle: Optional[str] = None
    status: Optional[str] = None
    icon: Optional[str] = None
    route: str
    updatedAt: Optional[str] = None
    privacyClass: str
    availability: EntityAvailability = "ok"


class RelatedEntity(BaseModel):
    ref: EntityRef
    relationship: str
    source: RelatedEntitySource = "fk"


class RelatedEntityItem(BaseModel):
    edge: RelatedEntity
    summary: EntitySummary


class EntityResolveRequest(BaseModel):
    refs: List[EntityRef] = Field(default_factory=list)


class EntityResolveResponse(BaseModel):
    items: List[EntitySummary] = Field(default_factory=list)


class EntityRelatedResponse(BaseModel):
    items: List[RelatedEntityItem] = Field(default_factory=list)


class EntitySearchResponse(BaseModel):
    query: str
    items: List[EntitySummary] = Field(default_factory=list)
    privacy_blocked: bool = False


class EntityReferenceCreate(BaseModel):
    source: EntityRef
    target: EntityRef
    relationship: AuthoredRelationship = "references"
    provenance: AuthoredProvenance = "user"
    anchor_json: Optional[str] = None
    client_event_id: Optional[str] = None


class EntityReferenceSyncRequest(BaseModel):
    source: EntityRef
    targets: List[EntityRef] = Field(default_factory=list)
    provenance: AuthoredProvenance = "user"


class EntityReferenceRead(BaseModel):
    id: str
    source: EntityRef
    target: EntityRef
    relationship: str
    provenance: str
    anchor_json: Optional[str] = None
    client_event_id: Optional[str] = None
    created_at: Optional[str] = None


class EntityReferenceListResponse(BaseModel):
    items: List[EntityReferenceRead] = Field(default_factory=list)


def unavailable_summary(ref: EntityRef, availability: EntityAvailability) -> EntitySummary:
    title = "Unavailable" if availability == "forbidden" else "Deleted" if availability == "deleted" else "Unknown"
    return EntitySummary(
        ref=ref,
        title=title,
        route=entity_route(ref.type, ref.id),
        privacyClass=data_class_for_entity_type(ref.type),
        availability=availability,
    )
